drivers/sd/dwcmshc/dwcmshc.h

```diff
@@ -0,0 +1,992 @@
+#define MSHC_IDSTS_DMA_WR                           0x7
+#define MSHC_IDSTS_DESC_CLOSE                       0x8
+#define MSHC_IDSTS_EB_SHIFT                         10
+#define MSHC_IDSTS_EB_MASK                          (0x3 << 10)
+
+//
+// MSHC_IDSTS/IDINTEN [9:0] fields
+//
+#define MSHC_IDINT_AIS                              (0x1 << 9)
+#define MSHC_IDINT_NIS                              (0x1 << 8)
+#define MSHC_IDINT_CES                              (0x1 << 5)
+#define MSHC_IDINT_DUI                              (0x1 << 4)
+#define MSHC_IDINT_FBE                              (0x1 << 2)
+#define MSHC_IDINT_RI                               (0x1 << 1)
+#define MSHC_IDINT_TI                               (0x1 << 0)
+
+#define MSHC_IDINT_ALL                              (MSHC_IDINT_AIS | \
+                                                     MSHC_IDINT_NIS | \
+                                                     MSHC_IDINT_CES | \
+                                                     MSHC_IDINT_DUI | \
+                                                     MSHC_IDINT_FBE | \
+                                                     MSHC_IDINT_RI  | \
+                                                     MSHC_IDINT_TI)
+
+//
+// MSHC_CARDTHRCTL fields
+//
+#define MSHC_CARDTHRCTL_CARD_RD_THRES_SHIFT         16
+#define MSHC_CARDTHRCTL_CARD_RD_THRES_MASK          (0xfff << 16)
+#define MSHC_CARDTHRCTL_CARD_WR_THRES_EN            (0x1 << 2)
+#define MSHC_CARDTHRCTL_BUSY_CLR_INT_EN             (0x1 << 1)
+#define MSHC_CARDTHRCTL_CARD_RD_THRES_EN            (0x1 << 0)
+
+//
+// MSHC_BACKEND_POWER fields
+//
+#define MSHC_BACKEND_POWER_ENABLE(n)                (0x1 << (n))
+
+//
+// MSHC_EMMCDDR_REG fields
+//
+#define MSHC_EMMCDDR_REG_HALF_START_BIT             (0x1 << 0)
+
+//
+// MSHC_RDYINT_GEN fields
+//
+#define MSHC_RDYINT_GEN_CNT_FINISH                  (0x1 << 24)
+#define MSHC_RDYINT_GEN_CNT_STATUS_SHIFT            16
+#define MSHC_RDYINT_GEN_CNT_STATUS_MASK             (0xff << 16)
+#define MSHC_RDYINT_GEN_WORKING                     (0x1 << 8)
+#define MSHC_RDYINT_GEN_MAXVAL_SHIFT                0
+#define MSHC_RDYINT_GEN_MAXVAL_MASK                 (0xff << 0)
+
+//
+// MSHC IDMAC fields
+//
+#define MSHC_IDMAC_DES0_OWN                         (0x1U << 31)
+#define MSHC_IDMAC_DES0_CES                         (0x1 << 30)
+#define MSHC_IDMAC_DES0_ER                          (0x1 << 5)
+#define MSHC_IDMAC_DES0_CH                          (0x1 << 4)
+#define MSHC_IDMAC_DES0_FS                          (0x1 << 3)
+#define MSHC_IDMAC_DES0_LD                          (0x1 << 2)
+#define MSHC_IDMAC_DES0_DIC                         (0x1 << 1)
+
+#define MSHC_IDMAC_DES1_BS2_SHIFT                   13
+#define MSHC_IDMAC_DES1_BS2_MASK                    (0x1fff << 13)
+#define MSHC_IDMAC_DES1_BS1_SHIFT                   0
+#define MSHC_IDMAC_DES1_BS1_MASK                    (0x1fff << 0)
+
+#pragma pack(1)
+
+typedef struct _MSHC_IDMAC_DESCRIPTOR {
```

Comment:
FWIW, current "official" Microsoft opinion of `typedef struct _STRUCT {...} STRUCT, *PSTRUCT;` is "strongly discourage".

- For code that needs to work on C, the recommendation is `typedef struct STRUCT {...} STRUCT;`.
- For code that only needs to be C++, the recommendation is `struct STRUCT {...};`.
- `_STRUCT` was just because the older MSVC compilers acted funny when both identifiers were the same, but this no longer seems to have any real benefit, and it conflicts with the C/C++ rule "only the compiler-provided headers should define stuff starting with double-underscore or underscore-capital".
- Most of the time, `PSTRUCT` has no advantage over `STRUCT*`, and it has several disadvantages (you have to look for two different types, and it discourages use of `const`).

Reply:
If you feel strongly about this, I can change it. I'm not a fan of PSTRUCT either, but nearly all existing MSFT driver code is written this way, and so I've been doing it just for consistency sake.

Reply:
I don't feel strongly. You do what seems best. Consistency is a good thing too.